CSS @imports in HTML missing semi-colon and space are mistakenly hidden from CSSPreload Scanner

Ahmad-S792 opened the pull request:
#### af134e645543841f47f56df5620d9fb742ef9426
<pre>
CSS @imports in HTML missing semi-colon and space are mistakenly hidden from CSSPreload Scanner

<a href="https://bugs.webkit.org/show_bug.cgi?id=253462">https://bugs.webkit.org/show_bug.cgi?id=253462</a>
rdar://problem/106666050

Reviewed by Ryosuke Niwa.

This patch aligns WebKit with Gecko / Firefox and Blink / Chromium.

Merge: <a href="https://chromium.googlesource.com/chromium/src.git/+/60092ea1a78a62ab84983e321b95415a358b07eb">https://chromium.googlesource.com/chromium/src.git/+/60092ea1a78a62ab84983e321b95415a358b07eb</a>

When scanning for @import rules to preload, CSSPreloadScanner terminates
a URL at &apos;;&apos; unconditionally. This is wrong as there might be &apos;;&apos; in a
URL, in which case we make a preload of a broken URL.

This patch scans URLs in a saner way:
- For urls given as a raw string, terminate at the ending quote
- Other those wrapped in a &apos;url()&apos;, terminate at the closing parenthesis

Note: We do not intend to write a full parser here, so it&apos;s still broken
in some cases (e.g., when the url contains &apos; &apos; or &apos;)&apos;). It&apos;s out of the
scope of this patch to fix them and already broken in current parser.

* Source/WebCore/html/parser/CSSPreloadScanner.cpp:
(CSSPreloadScanner::scan):
(CSSPreloadScanner::tokenize):
(CSSPreloadScanner::hasFinishedRuleValue):
* Source/WebCore/html/parser/CSSPreloadScanner.h: New bool function &apos;hasFinishedRuleValue&apos;
* LayoutTests/imported/w3c/web-platform-tests/loading/preloader-css-import-no-space.tentative.html: Add Test Case
* LayoutTests/imported/w3c/web-platform-tests/loading/preloader-css-import-no-space.tentative-expected.txt: Add Test Case Expectation
* LayoutTests/imported/w3c/web-platform-tests/loading/preloader-css-import-no-semicolon.tentative.html: Add Test Case
* LayoutTests/imported/w3c/web-platform-tests/loading/preloader-css-import-no-semicolon.tentative-expected.txt: Add Test Case Expectation
* LayoutTests/platform/glib/TestExpectations: Add Platform Specific Expectation (for GTK-WK2) with bug reference due to &apos;libsoup&apos; issue
* LayoutTests/platform/wpe/TestExpectations: Add Platform Specific Expectation (for WPE-WK2) with bug reference due to &apos;libsoup&apos; issue

Canonical link: <a href="https://commits.webkit.org/267322@main">https://commits.webkit.org/267322@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7432b6601ec0ad804d19b58bb7645fbcb8286a65

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/16286 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/16605 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/17024 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/18054 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/15273 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/16474 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/19685 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/16718 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/17653 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/16480 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/16923 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/13942 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/18821 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/14168 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/14742 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/21557 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/15155 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/14907 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/18209 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/15500 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/13135 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/14720 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/14583 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3894 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/19085 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/15327 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mcatanzaro commented on LayoutTests/platform/glib/TestExpectations:
```diff
@@ -1455,6 +1455,9 @@ imported/w3c/web-platform-tests/fetch/private-network-access/resources/xhr-sende
 imported/w3c/web-platform-tests/fetch/private-network-access/shared-worker.window.html [ DumpJSConsoleLogInStdErr ]
+webkit.org/b/260731 imported/w3c/web-platform-tests/loading/preloader-css-import-no-quote.tentative.html [ Skip ]
+webkit.org/b/260731 imported/w3c/web-platform-tests/loading/preloader-css-import-no-semicolon.tentative.html [ Skip ]
+webkit.org/b/260731 imported/w3c/web-platform-tests/loading/preloader-css-import-no-space.tentative.html [ Skip ]
```
Please mark them as `[ Failure ]`, not `[ Skip ]`. If you skip them, we'll never notice when they're fixed.

mcatanzaro replied:
```suggestion
webkit.org/b/260731 imported/w3c/web-platform-tests/loading/preloader-css-import-no-quote.tentative.html [ Failure ]
webkit.org/b/260731 imported/w3c/web-platform-tests/loading/preloader-css-import-no-semicolon.tentative.html [ Failure ]
webkit.org/b/260731 imported/w3c/web-platform-tests/loading/preloader-css-import-no-space.tentative.html [ Failure ]
```

Ahmad-S792 replied:
Will do in morning - 1 AM here. 😅

Ahmad-S792 replied:
Done. :-)

mcatanzaro commented on LayoutTests/platform/wpe/TestExpectations:
```diff
@@ -173,6 +173,9 @@ imported/w3c/web-platform-tests/webaudio/the-audio-api/the-destinationnode-inter
 http/tests/xmlhttprequest/state-after-network-error.html [ DumpJSConsoleLogInStdErr ]
+webkit.org/b/260731 imported/w3c/web-platform-tests/loading/preloader-css-import-no-quote.tentative.html [ Skip ]
+webkit.org/b/260731 imported/w3c/web-platform-tests/loading/preloader-css-import-no-semicolon.tentative.html [ Skip ]
+webkit.org/b/260731 imported/w3c/web-platform-tests/loading/preloader-css-import-no-space.tentative.html [ Skip ]
```
```suggestion
webkit.org/b/260731 imported/w3c/web-platform-tests/loading/preloader-css-import-no-quote.tentative.html [ Failure ]
webkit.org/b/260731 imported/w3c/web-platform-tests/loading/preloader-css-import-no-semicolon.tentative.html [ Failure ]
webkit.org/b/260731 imported/w3c/web-platform-tests/loading/preloader-css-import-no-space.tentative.html [ Failure ]
```

Ahmad-S792 replied:
Done. :-)

Ahmad-S792 committed:
CSS @imports in HTML missing semi-colon and space are mistakenly hidden from CSSPreload Scanner

https://bugs.webkit.org/show_bug.cgi?id=253462
rdar://problem/106666050

Reviewed by Ryosuke Niwa.

This patch aligns WebKit with Gecko / Firefox and Blink / Chromium.

Merge: https://chromium.googlesource.com/chromium/src.git/+/60092ea1a78a62ab84983e321b95415a358b07eb

When scanning for @import rules to preload, CSSPreloadScanner terminates
a URL at ';' unconditionally. This is wrong as there might be ';' in a
URL, in which case we make a preload of a broken URL.

This patch scans URLs in a saner way:
- For urls given as a raw string, terminate at the ending quote
- Other those wrapped in a 'url()', terminate at the closing parenthesis

Note: We do not intend to write a full parser here, so it's still broken
in some cases (e.g., when the url contains ' ' or ')'). It's out of the
scope of this patch to fix them and already broken in current parser.

* Source/WebCore/html/parser/CSSPreloadScanner.cpp:
(CSSPreloadScanner::scan):
(CSSPreloadScanner::tokenize):
(CSSPreloadScanner::hasFinishedRuleValue):
* Source/WebCore/html/parser/CSSPreloadScanner.h: New bool function 'hasFinishedRuleValue'
* LayoutTests/imported/w3c/web-platform-tests/loading/preloader-css-import-no-space.tentative.html: Add Test Case
* LayoutTests/imported/w3c/web-platform-tests/loading/preloader-css-import-no-space.tentative-expected.txt: Add Test Case Expectation
* LayoutTests/imported/w3c/web-platform-tests/loading/preloader-css-import-no-semicolon.tentative.html: Add Test Case
* LayoutTests/imported/w3c/web-platform-tests/loading/preloader-css-import-no-semicolon.tentative-expected.txt: Add Test Case Expectation
* LayoutTests/platform/glib/TestExpectations: Add Platform Specific Expectation (for GTK-WK2) with bug reference due to 'libsoup' issue
* LayoutTests/platform/wpe/TestExpectations: Add Platform Specific Expectation (for WPE-WK2) with bug reference due to 'libsoup' issue

Canonical link: https://commits.webkit.org/267322@main
